packages/grid/_modules_/grid/components/column-header-item.tsx

```diff
@@ -8,131 +8,132 @@ import { classnames } from '../utils';
-  onColumnDragStart?: (col: ColDef, currentTarget: HTMLElement) => void;
   separatorProps: React.HTMLAttributes<HTMLDivElement>;
 }
 
-export const ColumnHeaderItem = React.memo(
-  ({
-    column,
-    colIndex,
-    onColumnDragStart,
-    onColumnDragEnter,
-    onColumnDragOver,
-    isResizing,
-    separatorProps,
-    sortModel,
-    options,
-  }: ColumnHeaderItemProps) => {
-    const apiRef = React.useContext(ApiContext);
-    const { disableColumnReorder, showColumnRightBorder, disableColumnResize } = options;
+export const ColumnHeaderItem = ({
+  column,
+  isDragging,
+  colIndex,
+  isResizing,
+  separatorProps,
+  sortModel,
+  options,
+}: ColumnHeaderItemProps) => {
+  const apiRef = React.useContext(ApiContext);
+  const { disableColumnReorder, showColumnRightBorder, disableColumnResize } = options;
 
-    const columnSortModel = React.useMemo(
-      () =>
-        sortModel
-          .filter((model) => model.field === column.field)
-          .map((item) => ({
-            sortDirection: item.sort,
-            ...(sortModel.length <= 1 ? {} : { sortIndex: sortModel.indexOf(item) + 1 }),
-          }))[0] || {},
-      [column.field, sortModel],
-    );
+  const columnSortModel = React.useMemo(
+    () =>
+      sortModel
+        .filter((model) => model.field === column.field)
+        .map((item) => ({
+          sortDirection: item.sort,
+          ...(sortModel.length <= 1 ? {} : { sortIndex: sortModel.indexOf(item) + 1 }),
+        }))[0] || {},
+    [column.field, sortModel],
+  );
 
-    let headerComponent: React.ReactElement | null = null;
-    if (column.renderHeader) {
-      headerComponent = column.renderHeader({
-        api: apiRef!.current!,
-        colDef: column,
-        colIndex,
-        field: column.field,
-      });
-    }
+  let headerComponent: React.ReactElement | null = null;
+  if (column.renderHeader) {
+    headerComponent = column.renderHeader({
+      api: apiRef!.current!,
+      colDef: column,
+      colIndex,
+      field: column.field,
+    });
+  }
 
-    const dragConfig = {
-      draggable:
-        !disableColumnReorder && !!onColumnDragStart && !!onColumnDragEnter && !!onColumnDragOver,
-      onDragStart: onColumnDragStart && ((event) => onColumnDragStart(column, event.currentTarget)),
-      onDragEnter: onColumnDragEnter && ((event) => onColumnDragEnter(event)),
-      onDragOver:
-        onColumnDragOver &&
-        ((event) => {
-          onColumnDragOver(column, {
-            x: event.clientX,
-            y: event.clientY,
-          });
-        }),
-    };
-    const width = column.width!;
+  const dragConfig = {
+    draggable:
+      !disableColumnReorder &&
+      !!apiRef!.current.onColItemDragStart &&
+      !!apiRef!.current.onColItemDragEnter &&
+      !!apiRef!.current.onColItemDragOver,
+    onDragStart:
+      apiRef!.current.onColItemDragStart &&
+      ((event) => apiRef!.current.onColItemDragStart(column, event.currentTarget)),
```

Comment:
I would refactor this block like that, 
- remove the dragConfig object
then define each function with useCallback ie
```
const onDragStart = React.useCallback(event =>  {
apiRef!.current.onColItemDragStart(column, event.currentTarget));
} , [column]);
...
```

then
```
const dragConfig = { onDragStart, onDragEnd, ...}
render: (
 <div className="MuiDataGrid-colCell-draggable" {...dragConfig}>
```

Reply:
I think that we are getting too far with useCallback. What if we forbid it for handlers that are applied on primitives and aren't called repeatively. For instance, it's important to have useCallback for a mouse move but for a mouse start, this creat friction in development without improving performance (it could very well harm it).

Reply:
if you memo everything and leave a drag handler out then, your underlying component will rerender and hit perf for all the other optimised prop, this is due to a new function that will be defined at every render. So in case of streaming when you render super fast, it can be a perf hit...

The goal is to write good optimised code. 

And by definition, useCallback is here to define a function, and store it for the next render so it doesn't need to redefined it unless the dependencies changed. Hence we use it everywhere. 
It's a bit like instance method stored on the prototype vs method created in the constructor of a function, redefined for each instance.

Reply:
⚠️ We are talking about a primitive here: `<div>` which doesn't implement any memorization.

Reply:
> don't optimize your code unless you can measure it. 

Not in favor of that, as to me, it means, writes bad unoptimized code everywhere or don't follow any react practices, then when it gets bloated or you start to notice some perf issues, waste time profiling everything, refactor everything, and finally when you wasted a lot of time and headache, adopt some react practices.

Reply:
if nothing changed, divs won't rerender

Reply:
Haha, that's definitely a valid way to see things! My experience has been the opposite: write code you think is optimized. Later on, notice a performance issue. Dive into understanding what's going on, realize that your optimizations are harming or doing nothing, remove them 😆.

Reply:
Well we are talking about less than 1ms so it will be difficult to measure. But if we multiply it by 10000000, with very fast updates... then you will see limitations...

I'd rather write super optimised code and if issues investigate and remove, instead of the opposite. 
As there are less chances of making mistakes that way

Reply:
This is an interesting discussion. I don't think we need it that badly here firstly because the check is not really a heavy computation, and secondly because the underlying method (for instance the `onColItemDragStart `) is already a memorized function. I don't want to go against the grain as I am the least experienced here on that codebase but I'm also against optimizing for the sake of optimizing. I'm not saying that we should write shitty code but if there are performance problems down the road we can analyze and solve them later.

Reply:
No objection to Damien's direction if he has a strong perspective on this. I would simply add that to my experience that memorization of handlers is only important in two cases:

- repeatedly fired events, if you don't memorize you will lose important events (React attach and detach at each render). In our case, it can be the mouse move, our a drag over events, or dependency of useEffect.
- high in the tree React component that computation is heavy and that you can prune with memorization. In our case, it can be the `Viewport` component (because it's large) or a Cell (because there are many)

in other cases, it doesn't matter.

Reply:
Updated